### PR TITLE
Store the status of a step by step

### PR DESCRIPTION
**Pull request by leenagupte**
Trello: https://trello.com/c/0H0pphSZ
Implements: PR #714 

## What's changed?
Creates a new field to store the status of a step by step. The field is validated against the current list of possible statuses:

- draft
- published
- scheduled

It also removes "unpublished_changes" as a status as, in terms of workflow, this is the same as being in "draft".

The "live" status has also been renamed to "published" to to be consistent with publishing-api the other publishing apps.

A rake task has been created to populate the current status of all existing step by steps.

## Why
The "status" of a step by step was calculated from data about the step by step, e.g. if there is a draft_updated_at date but not a published_at date, then it's status was a draft etc.

That worked until now, when we only had 4 statuses:

- Draft
- Live (published)
- Scheduled
- Unpublished changes

To add the 2i workflow, we'll need at least three more statuses:

- Submitted for 2i
- In review
- 2i approved

Which will make calculating them very complicated.

## Post deployment task

Run task: `rake step_by_step:set_status`

### CODE REVIEW

File: app/models/step_by_step_page.rb [hunk 1]
```diff
@@ -15,6 +15,7 @@ class StepByStepPage < ApplicationRecord
   validates :slug, format: { with: /\A([a-z0-9]+-)*[a-z0-9]+\z/ }, uniqueness: true
   validates :slug, slug: true, on: :create
   validates :status, inclusion: { in: STATUSES }, presence: true
+  validates :status, status: true
```
Comment (erino) [inline app/models/step_by_step_page.rb]: What about renaming the validator as it took me quite a while to realise that the second `:status` was the name of the validator?

File: app/models/step_by_step_page.rb [hunk 2]
```diff
@@ -30,7 +30,7 @@ def has_draft?
   end
 
   def scheduled_for_publishing?
-    has_draft? && scheduled_at.present?
+    has_draft? && scheduled_at.present? && status == "scheduled"
```
Comment (erino) [inline app/models/step_by_step_page.rb]: What about using a [`StringInquirer`](https://api.rubyonrails.org/v5.2/classes/ActiveSupport/StringInquirer.html) to turn status into methods? such as `status.scheduled?`

File: app/presenters/step_by_step_page_presenter.rb [hunk 3]
```diff
@@ -9,7 +9,7 @@ def initialize(step_by_step_page)
 
   def summary_metadata
     items = {
-      "Status" => step_by_step_page.status[:text],
+      "Status" => step_by_step_page.status.humanize,
```
Comment (erino) [inline app/presenters/step_by_step_page_presenter.rb]: Is the capitalisation of `unpublished_changes` to "Unpublished changes" correct? or should it be "Unpublished Changes"

File: app/views/step_by_step_pages/index.html.erb [hunk 4]
```diff
@@ -132,7 +132,7 @@
               <td class="govuk-table__cell">
                 <% if step_by_step_page.has_draft? %>
                   <%= link_to "/#{step_by_step_page.slug}", preview_url(step_by_step_page.slug, auth_bypass_id: step_by_step_page.auth_bypass_id), target: "_blank" %>
-                <% elsif step_by_step_page.status == "live" %>
+                <% elsif step_by_step_page.status == "published" %>
                   <%= link_to "/#{step_by_step_page.slug}", live_url(step_by_step_page.slug), target: "_blank" %>
```
Comment (erino) [inline app/views/step_by_step_pages/index.html.erb]: What about changing `live_url` to `public_url`? as `live` is being purged from the codebase.

File: app/views/step_by_step_pages/index.html.erb [hunk 5]
```diff
@@ -51,35 +51,7 @@
...
-              },
-              {
-                text: "Scheduled",
-                value: "scheduled",
-                data_attributes: {
-                  show: "scheduled"
-                },
-              }
-            ].each{ |item| item[:selected] = item[:value] == params[:status] }
+            options: StepByStepFilter::Options.statuses.each{ |item| item[:selected] = item[:value] == params[:status] }
```
Comment (ChrisBAshton) [inline app/views/step_by_step_pages/index.html.erb]: Nice!

Think it should have a space between `each` and `{`.

File: spec/validators/status_validator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      step_by_step_page.scheduled_at = nil
+      step_by_step_page.status = "scheduled"
+
+      expect(step_by_step_page).not_to be_valid
+      expect(step_by_step_page.errors.messages[:status]).to eq([error_message])
+    end
+
+    it "does not allow status to be scheduled if there is no draft" do
+      allow(step_by_step_page).to receive(:has_draft?).and_return(false)
+      step_by_step_page.scheduled_at = Time.zone.now + 1.hour
```
Comment (erino) [inline spec/validators/status_validator_spec.rb]: `1.hour.since` is a bit more Railsy

File: app/models/step_by_step_page.rb [hunk 7]
```diff
@@ -49,6 +49,7 @@ def mark_as_published
     update_attribute(:draft_updated_at, now)
     update_attribute(:scheduled_at, nil)
     update_attribute(:assigned_to, nil)
+    update_attribute(:status, "published")
```
Comment (erino) [inline app/models/step_by_step_page.rb]: Can all of these (and others from other methods) calls to `update_attribute` be rewritten one call to `update { published_at: now, draft_updated_at: now ... }`  saving a few database calls

File: app/services/step_by_step_page_reverter.rb [hunk 8]
```diff
@@ -12,6 +12,7 @@ def repopulate_from_publishing_api
     step_by_step_page.introduction = introduction
     step_by_step_page.description = payload_from_publishing_api[:description]
     step_by_step_page.draft_updated_at = step_by_step_page.published_at
+    step_by_step_page.status = "published"
```
Comment (erino) [inline app/services/step_by_step_page_reverter.rb]: Can all of these updates be replaced with a single call to `update`?

File: app/views/step_by_step_pages/index.html.erb (hunk 5 above)
Comment (erino) [reply app/views/step_by_step_pages/index.html.erb]: and what moving this to a method in `StepByStepFilter::Options` to keep the view cleaner?

Comment (leenagupte) [reply app/views/step_by_step_pages/index.html.erb]: @erino That's an interesting idea. I'll investigate how to do that.

File: app/presenters/step_by_step_page_presenter.rb (hunk 3 above)
Comment (leenagupte) [reply app/presenters/step_by_step_page_presenter.rb]: The [original text](https://github.com/alphagov/collections-publisher/blob/master/app/views/step_by_step_pages/index.html.erb#L83) is "Unpublished changes", so I assumed that's it's alright.

File: app/services/step_by_step_page_reverter.rb (hunk 8 above)
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: Isn't that what happens when `save!` is called?

File: app/models/step_by_step_page.rb [hunk 9]
```diff
@@ -53,7 +53,7 @@ def self.validate_redirect(redirect_url)
 
   def status
     return "scheduled" if scheduled_for_publishing?
-    return "unpublished_changes" if unpublished_changes?
+    return "draft" if has_draft?
```
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/models/step_by_step_page.rb]: This shouldn't be a problem as it should be picked up by the tests.

File: spec/services/step_by_step_filter_spec.rb [hunk 10]
```diff
@@ -40,16 +32,6 @@
...
 
-    it "returns a list of step by steps with a status of published" do
-      filter_params = {
-        status: "published"
-      }
-      results = described_class.new(filter_params).results
-
-      expect(results.count).to eq(1)
-      expect(results.first.title).to eq(published_step_by_step.title)
-    end
```
[the remark this answers is not in the document]
Comment (leenagupte) [reply spec/services/step_by_step_filter_spec.rb]: > The Filter specs were using `unpublished_changes` to check for what happens when filtering by a
status with nothing assigned. This has been changed to use `published` instead as that required the
least amount of test changes.

To do that `published_step_by_step_page` had to be removed. So this test will never pass.

File: app/models/step_by_step_page.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/models/step_by_step_page.rb]: I do that in a later commit.

File: app/models/step_by_step_page.rb [hunk 11]
```diff
@@ -52,6 +52,10 @@ def mark_as_published
   def mark_as_unpublished
     update_attribute(:published_at, nil)
     update_attribute(:draft_updated_at, nil)
+
+  def mark_as_scheduled
+    update_attribute(:status, "scheduled")
+  end
```
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/models/step_by_step_page.rb]: Ah, the `end` for `mark_as_unpublished` is in the wrong place. Will fix.

File: app/presenters/step_by_step_page_presenter.rb (hunk 3 above)
Comment (erino) [reply app/presenters/step_by_step_page_presenter.rb]: 👍

File: app/services/step_by_step_page_reverter.rb (hunk 8 above)
Comment (erino) [reply app/services/step_by_step_page_reverter.rb]: I mean, could the entire methods be rewritten like this?

```ruby 
def repopulate_from_publishing_api
  update(
    title: payload_from_publishing_api[:title],
    slug: payload_from_publishing_api[:base_path].tr('/', ''),
    introduction: introduction,
    description: payload_from_publishing_api[:description],
    draft_updated_at: step_by_step_page.published_at,
    status: "published"
  )
end
```

File: spec/services/step_by_step_page_reverter_spec.rb [hunk 12]
```diff
@@ -68,6 +68,10 @@
       expect(step_by_step.draft_updated_at).to eq(published_at)
     end
 
+    it "sets the status to published" do
+      expect(step_by_step_page.status).to eq("published")
```
Comment (erino) [inline spec/services/step_by_step_page_reverter_spec.rb]: If `StringInquirer` is used for status, then I think this spec could use a predicate matcher

```ruby
expect(step_by_step_page.status).to be_published
```

File: app/models/step_by_step_page.rb (hunk 1 above)
Comment (leenagupte) [reply app/models/step_by_step_page.rb]: I've renamed the validator to `StatusPrerequisiteValidator`.

File: app/models/step_by_step_page.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/models/step_by_step_page.rb]: Updated and fixed up.

File: app/models/step_by_step_page.rb (hunk 1 above)
Comment (leenagupte) [reply app/models/step_by_step_page.rb]: Updated and fixed-up

File: app/models/step_by_step_page.rb (hunk 2 above)
Comment (leenagupte) [reply app/models/step_by_step_page.rb]: Updated and fixed-up

File: app/views/step_by_step_pages/index.html.erb (hunk 4 above)
Comment (leenagupte) [reply app/views/step_by_step_pages/index.html.erb]: Updated and fixed-up.

File: app/models/step_by_step_page.rb (hunk 7 above)
Comment (leenagupte) [reply app/models/step_by_step_page.rb]: Updated and fixed-up

File: spec/validators/status_validator_spec.rb (hunk 6 above)
Comment (leenagupte) [reply spec/validators/status_validator_spec.rb]: Updated and fixed-up

File: app/services/step_by_step_page_reverter.rb (hunk 8 above)
Comment (leenagupte) [reply app/services/step_by_step_page_reverter.rb]: Update and fixed-up

File: spec/services/step_by_step_page_reverter_spec.rb (hunk 12 above)
Comment (leenagupte) [reply spec/services/step_by_step_page_reverter_spec.rb]: Update and fixed-up

File: app/views/step_by_step_pages/index.html.erb (hunk 5 above)
Comment (leenagupte) [reply app/views/step_by_step_pages/index.html.erb]: @erino Refactored and fixed-up

**Pull request merged**